### DISCUSSION TITLE
Szurubooru backend rewrite

### DISCUSSION THREAD

**Discussion by liamw1** [rr-/szurubooru#705] [Show and tell]
About six months ago I started working on a total rewrite of the backend on my [fork](https://github.com/liamw1/oxibooru) with a focus on improving performance. It's finally far enough along that I'm comfortable distributing a first release. I wouldn't say it's production ready, but it should be fine for personal use. 

I've made a number of changes and improvements to the backend:

- Most queries are much faster (some 60x faster in testing) and scale better with larger databases
- Near-instant startup time
- Reverse image search is slightly better at discerning similar images from dissimilar ones
- Reverse image search is faster and more thorough
- Improved tag implication/suggestion cyclic dependency detection
- Post navigation via neighbors respects sorting
- Ability to sort posts by pool count, uploader, aspect ratio, and safety
- Ability to filter tags that have a specific implication or suggestion
- Server no longer locks up when mass-deleting related posts
- Separate secrets for hashing passwords and post ids (#580)

If you're interested in trying it out, I've also created a script to convert a Szurubooru database into one compatible with my fork. Though this has some limitations which are noted in the [conversion guide](https://github.com/liamw1/oxibooru/blob/master/doc/CONVERSION.md).

Questions and contributions are welcome.

**Comment by neobooru** [comment]
Looks very cool, though I haven't fully explored it yet. Are the known limitations in https://github.com/liamw1/oxibooru/blob/master/doc/CONVERSION.md the only missing/different features, or are there other not-yet-implemented features?

**Reply by liamw1** [reply]
That basically covers it. The [API](https://github.com/rr-/szurubooru/blob/master/doc/API.md) is almost fully implemented; the only unimplemented feature is [listing snapshots](https://github.com/rr-/szurubooru/blob/master/doc/API.md#listing-snapshots). Password reset requests are technically implemented but are unlikely to actually work because they are entirely untested (I haven't set up a SMTP server to test them yet).

**Reply by liamw1** [reply]
Update: The API has now been more or less fully implemented. Password reset is functional and snapshots have been added to the backend. Additionally, two of the limitations of the conversion process have been addressed. Those being post notes and snapshots; both of which are now handled by the conversion script.

**Comment by bakedhorse** [comment]
Question about oxibooru. Is it going to be a complete 1:1 rust port of szurubooru, and stay like that, or will there be added features later on such as descriptions and/or block lists?

**Reply by liamw1** [reply]
I'm willing to add new features if people want them. Though, I'm limited in my ability to implement things in the frontend because I don't understand the client code as well (and also I don't know JS). There are already minor features added in Oxibooru such as new filters for tags and posts.
